### PR TITLE
chore: 5967 validate metric naming rules

### LINKED ISSUE #5967: Enforce metric name rules
There are currently no restrictions on what characters can be used in metric names. Some characters cause problems for systems like Prometheus and Elastic to ingest. There are currently scripts that modify metric names to make them compatible with other systems.

Metric names should be validated at the time of creation. There are at least two options for this:

1. Throw an exception when a metric is created with an invalid name
2. Modify the name of the metric to conform to the standards when publishing to externals 

**Elastic converter:** https://github.com/swirlds/infrastructure/blob/master/hedera-utils/PipelineGenerator.py
**Prometheus converter:** `com.swirlds.common.metrics.platform.prometheus.NameConverter`

### PR DESCRIPTION
**Pull request by mxtartaglia-sl**
**Description**:
Since now, developer provided metric names, categories, and units are checked against offending Prometheus naming limitations listed here: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels, in case of finding an offending naming an error level message is logged with an EXCEPTION marker that will cause the JRS test to fail. Still, as it was being done, the matric name gets modified in order to be able to expose the metric regardless.
Unit field has been incorated to the check.   

**Related issue(s)**:

Fixes #5967

### CODE REVIEW

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:60]: At first glance this looked like a no-op because the call on line 59 would create the `values` object using the given `metric` object already. However, I see that the implementation of the `CommonMetricValues` constructor actually calls `fix()` on each of the individual values, while the `checkAgainst` uses a simple Objects.equals() to compare the individual values.

So essentially, in these two lines of code we're trying to assert that `fix()` returns its argument verbatim, or else we log a bunch of errors.

This brings the question as to
1. whether this is the intended purpose of this code,
2. and/or whether the calls to `fix()` are required (because this check will fail the moment the `fix()` returns anything different from its input argument),
3. and if they are required, then shouldn't the checkAgainst() take the `fix()` into account?

Could you please clarify?

Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:60]: Sure, the "fix()" method is the existing mechanism to replace a metric name component (category/name/unit), so it is valid against some predefined naming rules Prometheus requires. If the result of invoking fix() for a string returns the exact string, the value is a valid name for Prometheus.   
The requirement is that we want to avoid/identify when the metric name is changed under the hood by the framework and inform developers that they should adjust the metric name. That is why it compares the actual parameter vs. the value after being #fix()ed.
Let me rethink the design so it is more explicit; I see how having a check method that only performs log statements collaborates with the idea that something must be wrong.
The initial design had all the code in the constructor and exposed the values after being fixed to the subclasses, but that seemed weak encapsulation.

Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:60]: Got it. Thanks for the clarification. This makes sense. I'd suggest to add a code comment (either in checkAgainst() javadoc, or above the call to the method) to briefly describe the intention of the check. This would make the code a lot clearer.

Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java]: @anthony-swirldslabs Let me know if the logic + name change + additional javadoc did a better job expressing the intention

Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java]: That looks perfect! Thank you!

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:104]: I do not get that flag. Metric defines the type field. Why not just check if that field is null?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/MeasuredBlockingQueue.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/MeasuredBlockingQueue.java:37]: Will this change metrics names that are current used?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/QueueThreadMetrics.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/QueueThreadMetrics.java:59]: Will this change metrics names that are current used?

File: platform-sdk/swirlds-metrics-api/src/main/java/com/swirlds/metrics/api/Metrics.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-metrics-api/src/main/java/com/swirlds/metrics/api/Metrics.java:39]: Will this change metrics names that are current used?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java:78]: This does not only change the metrics name. I do not know if platform uses the name somehow else. cc @cody-littley @lpetrovic05

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java:64]: We really need to check after the merge that all the grafana dashboards still work

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:104]: The problem is that prometheus client will fail if we set a unit to a metric that should not have a unit set.
But our code doesn't prevent this from happening; we allow users to set units in any metric type even when Prometheus fails (we have one occurrence that sets "boolean" and other values in tests), which makes it impossible to rely only on a check against null. What was happening before is that the adapters knew not to set the unit field when necessary. I needed that information in the superclass instead of each adapter.
I guess that the final fix would be not to allow to set units on invalid metrics, and we could remove the flag, just that the change was way bigger.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/MeasuredBlockingQueue.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/MeasuredBlockingQueue.java:37]: It will not change the name on the Prometheus side; they were already called this after the adapter fixed the name.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/QueueThreadMetrics.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/framework/internal/QueueThreadMetrics.java:59]: It will not change the name on the Prometheus side; they were already called this after the adapter fixed the name.

File: platform-sdk/swirlds-metrics-api/src/main/java/com/swirlds/metrics/api/Metrics.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-metrics-api/src/main/java/com/swirlds/metrics/api/Metrics.java:39]: It will not change the name on the Prometheus side; they were already called this after the adapter fixed the name.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java:78]: I checked with @poulok before pushing the change... the naming is less readable but at least consistent with the metrics created under the hood.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java:64]: True that, although I haven't found any difference between the resulting metric names and the values in graphana json other than the title of the metric.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:62]: nullity annotation for return type

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:102]:
```suggestion
        private ConvertedMetricValues(final boolean unitless, final @NonNull Metric metric) {
```

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:112]: nullity annotation for return type

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:128]:
```suggestion
        private void verifyMetricNamingComponents(@NonNull final Metric metric) {
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/status/PlatformStatusManager.java:78]: Thread names are purely used for understanding thread dumps and for logging. There are no hard dependencies I'm aware of. Anyways, we are very soon going to completely drop support for this style of threads, so naming is not super important.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/uptime/UptimeMetrics.java:64]: From my perspective, it's not the end of the world if a metric name changes. Most of our current metric names are not so great, and I have intentions to change many of them. 😜

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (OlegMazurov) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:31]: `Logger logger` is predominantly used in platform code.

**Review by OlegMazurov — APPROVED** [review]

**Review by cody-littley — DISMISSED** [review]

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:31]: Thanks, @OlegMazurov  I should have searched what the pattern was. If it is ok with you, if there is any other suggestion that I should commit, I'll change the name in this PR, but if not, since platform-base is the owner of this file, I can create a follow-up task to replace the name that will require fewer approvals for merging it...

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (hendrikebbers) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:104]: ok, if I understood it correctly "supportsUnit" would be a better param from my point of view. That one would define if the metric supports a unit. Next to that the Javadoc is missing that would help to understand the param

Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java:104]: Agreed, `supportsUnit` sounds better. I had the intention that the default false value set the unit instead of not setting it, but `unitless` ended up being a strange name. 
I didn't add Javadoc in this method given that it is a private method on a private inner class, the inner class has a Javadoc, and the containing class also has one explaining the receiving parameters. I'll add one here too, but I'm checking to be on the same page for the future: Should we always add Javadoc or only in public interfaces?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java
Comment (mxtartaglia-sl) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java]: Soved

**Review by cody-littley — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by OlegMazurov — APPROVED** [review]

**Review by hendrikebbers — APPROVED** [review]

**Pull request merged**
